csp/impl/types/typing_utils.py

```diff
@@ -47,7 +50,163 @@ def get_orig_base(cls, typ):
             return numpy.ndarray
         return res
 
+    @classmethod
+    def pretty_typename(cls, typ):
+        if cls.is_generic_container(typ):
+            return str(typ)
+        elif cls.is_forward_ref(typ):
+            return cls.pretty_typename(typ.__forward_arg__)
+        elif isinstance(typ, type):
+            return typ.__name__
+        else:
+            return str(typ)
+
 
-# Current typing utilities were
-# stabilized as of python 3.7
 CspTypingUtils = CspTypingUtils37
+
+if sys.version_info >= (3, 9):
+
+    class CspTypingUtils39(CspTypingUtils37):
+        # To support PEP 585
+        _GENERIC_ALIASES = (typing._GenericAlias, typing.GenericAlias)
+
+    CspTypingUtils = CspTypingUtils39
+
+if sys.version_info >= (3, 10):
+
+    class CspTypingUtils310(CspTypingUtils39):
+        # To support PEP 604
+        @classmethod
+        def is_union_type(cls, typ):
+            return (isinstance(typ, typing._GenericAlias) and typ.__origin__ is typing.Union) or isinstance(
+                typ, types.UnionType
+            )
+
+    CspTypingUtils = CspTypingUtils310
+
+
+class TsTypeValidator:
+    """Class to help validate the arg of TsType.
+    For example, this is to make sure that:
+        ts[List] can validate as ts[List[float]]
+        ts[Dict[str, List[str]] won't validate as ts[Dict[str, List[float]]
+        ts["T"], ts[TypeVar("T")], ts[List["T"]], etc are allowed
+        ts[Optional[float]], ts[Union[float, int]], ts[Annotated[float, None]], etc are not allowed
+        etc
+    For validation of csp baskets, this piece becomes the bottleneck
+    """
+
+    _cache: typing.Dict[typing.Type, "TsTypeValidator"] = {}
+
+    @classmethod
+    def make_cached(cls, source_type: typing.Type):
```

Comment:
I missed the note on this, where do you see that typing.Type is deprecated?

Reply:
https://docs.python.org/3/library/typing.html#typing.Type

Reply:
Ah, it's deprecated in the same way that all the typing hints are deprecated, i.e. List, Dict, etc. Since we still support python 3.8, I suggest we just leave it as typing.Type, and when we drop 3.8 support, we can go change everything from typing to use the builtins at once.

Reply:
But we don't use it as a Generic (with an inner type) so I believe we can just leave it as `type` which works for all Python versions.

Reply:
I see your point, will fix.